subprojects/core/src/main/java/org/gradle/api/internal/tasks/options/InstanceOptionDescriptor.java

```diff
@@ -70,10 +71,15 @@ public Class<?> getArgumentType() {
         return optionElement.getOptionType();
     }
 
+    @SuppressWarnings({"rawtypes", "unchecked"})
     private List<String> readDynamicAvailableValues() {
         if (optionValueMethod != null) {
-            Collection values = optionValueMethod.invoke(object);
-            return CollectionUtils.toStringList(values);
+            Object values = optionValueMethod.invoke(object);
+            if (values instanceof Provider) {
+                return CollectionUtils.toStringList(((Provider<Collection>) values).get());
```

Comment:
🤔 Would this be more correct?

```suggestion
                return CollectionUtils.toStringList(((Provider<? extends Iterable<?>>) values).get());
```

Reply:
Since we allow only `Provider<Collection>` it make sense we cast to `Provider<Collection>>`